Runtimes/Unity/BepInEx.Unity.Mono.Preloader/UnityPreloader.cs

```diff
@@ -130,6 +137,61 @@ public static void Run()
         }
     }
 
+    // First step of the chainloader entrypoint: Harmony patch the main Unity assembly with a suitable hook
+    private static void OnAssemblyLoad(object sender, AssemblyLoadEventArgs args)
+    {
+        const string sceneManagerTypeName = "UnityEngine.SceneManagement.SceneManager";
+        const string displayTypeName = "UnityEngine.Display";
+        
+        var assembly = args.LoadedAssembly;
+        var assemblyName = assembly.GetName().Name;
+
+        if (assemblyName is not ("UnityEngine.CoreModule" or "UnityEngine"))
+            return;
+
+        AppDomain.CurrentDomain.AssemblyLoad -= OnAssemblyLoad;
+
+        var sceneManagerType = assembly.GetType(sceneManagerTypeName, false);
+        if (sceneManagerType != null)
+        {
+            var activeSceneChangedMethod = sceneManagerType.GetMethod("Internal_ActiveSceneChanged",
+                                                                      BindingFlags.NonPublic | BindingFlags.Static);
+            Harmony.Patch(activeSceneChangedMethod, prefix: new HarmonyMethod(typeof(UnityPreloader), nameof(Entrypoint)));
+            Log.LogInfo($"Hooked into {activeSceneChangedMethod.FullDescription()}");
+            return;
+        }
+
+        var displayType = assembly.GetType(displayTypeName, false);
+        if (displayType != null)
+        {
+            var recreateDisplayListMethod = displayType.GetMethod("RecreateDisplayList", BindingFlags.NonPublic | BindingFlags.Static);
+            Harmony.Patch(recreateDisplayListMethod, postfix: new HarmonyMethod(typeof(UnityPreloader), nameof(Entrypoint)));
+            Log.LogInfo($"Hooked into {recreateDisplayListMethod.FullDescription()}");
+            return;
+        }
+
+        Log.LogError($"Couldn't find a suitable chainloader entrypoint in the {assemblyName} assembly because " +
```

Comment:
This should be a Fatal log, also a trycatch wraparound would be a good idea to provide a better error message.

Reply:
Done and added a try catch with a different fatal error